radix-engine/src/kernel/call_frame.rs

```diff
@@ -759,35 +762,51 @@ impl<C, L: Clone> CallFrame<C, L> {
             }
         }
 
-        let mut handler = WrapperHandler {
+        let mut adapter = CallFrameToIOStoreAccessAdapter {
             call_frame: self,
             handler,
             phantom: PhantomData::default(),
         };
 
-        substate_io.create_node(&mut handler, destination_device, node_id, node_substates)?;
+        substate_io.create_node(destination_device, node_id, node_substates, &mut adapter)?;
 
         Ok(())
     }
 
     /// Removes node from call frame and owned nodes will be possessed by this call frame.
-    pub fn drop_node<S: CommitableSubstateStore>(
+    pub fn drop_node<'x, E: 'x, S: CommitableSubstateStore>(
         &mut self,
         substate_io: &mut SubstateIO<S>,
         node_id: &NodeId,
-    ) -> Result<NodeSubstates, DropNodeError> {
+        handler: &mut impl CallFrameStoreAccessHandler<C, L, E>,
+    ) -> Result<NodeSubstates, CallbackError<DropNodeError, E>> {
         self.take_node_internal(node_id)
-            .map_err(DropNodeError::TakeNodeError)?;
-
-        let node_substates = substate_io.drop_node(SubstateDevice::Heap, node_id)?;
-        let mut handler = NullHandler::new();
+            .map_err(|e| CallbackError::Error(DropNodeError::TakeNodeError(e)))?;
+
+        let (frame, handler, node_substates) = {
```

Comment:
I don't believe you need to pass back `frame` and `handler` from this block. You should just be able to use `self` and the original `handler` since `adapter` is scoped.

Reply:
I will give it a try :)